ch2 dart fwe (#6611)
dart-lang/site-www: src/content/get-started/ch2-add-commands.md

    * `arguments.isNotEmpty` checks if any command-line arguments were
        provided.
    * `arguments.first` accesses the very first argument, which we're using as
        our command.
    * Notice the `version` variable is declared as a `const`. This means its
        value is known at compile time and you cannot change it during runtime.
    * `arguments`, on the other hand, is a regular (non-constant) variable
        because its content can change based on user input.

    Test the `help` command: Run your application with the help argument:

    ```bash
    dart bin/cli.dart help
    ```

    You should see the usage information printed.

    Also, try running it without any arguments:

    ```bash
    dart bin/cli.dart
    ```

    It should now also display the usage information.

    :::note
    At this point, any command we haven't defined will also print usage
    information. This is expected behavior for now.
    :::

### Task 2: Implement the search command

Next, implement a basic `search` command that takes an article title as
input. As you build this functionality, you'll work with `List` manipulation,
null checks, and string interpolation.

> 1.  **Integrate the `search` command into `main`:** First, modify your `main`

Reviewer: ```suggestion
1.  **Integrate the `search` command into `main`:** First, modify the `main`
```

    function in `cli/bin/cli.dart` to include an `else if` branch that handles

>     the `search` command. For now, we'll just print a placeholder message.

Reviewer: ```suggestion
    the `search` command. For now, just print a placeholder message.
```

    ```dart
    void main(List<String> arguments) {
      if (arguments.isNotEmpty && arguments.first == 'version') {
        print('Dartpedia CLI version $version');
      } else if (arguments.isNotEmpty && arguments.first == 'help') {
        printUsage();
      } else if (arguments.isNotEmpty && arguments.first == 'search') {
        // Add this new block:
        print('Search command recognized!');
      } else {
        printUsage();
      }
    }
    ```

1.  **Test the new command:** Run your application with the `search`
    command:

    ```bash
    dart bin/cli.dart search
    ```

    You should see:

    ```bash
    Search command recognized!
    ```

1.  **Define the `runApp` function:** The `search` command will eventually run
    the core logic of your application by calling a function called `runApp`.

>     For now, it will print the arguments passed into search. Place this new

Reviewer: ```suggestion
    For now, have `runApp` print the arguments passed into it with the `search` command. Place this new
```

    function below `main`.

    ```dart
    // ... (your existing main function)

    void runApp(List<String>? arguments) { // Add this new function and add ? to arguments type
      print('runApp received arguments: $arguments');
    }

    // ... (your existing printUsage() function)
    ```

>     * The `List<String>? arguments` type signature means that the `arguments`

Reviewer: I think you should preface these bulleted lists with something like this:

Highlights from the preceding code:

        list itself can be `null`. This is important for null safety in Dart.

> 1.  **Call `runApp` from `main`:** Now, modify the `search` command block in

Reviewer: ```suggestion
1.  **Call the `runApp` function from the `main` function:** Now, modify the `search` command block in
```

    `main` to call `runApp` and pass it any arguments that come after the

>     `search` command itself. We use `arguments.sublist(1)` to get all

Reviewer: ```suggestion
    `search` command itself. Use `arguments.sublist(1)` to get all
```

    arguments starting from the second one. If no arguments are provided after

>     `search`, we'll pass `null` to `runApp`.

Reviewer: ```suggestion
    `search`, pass `null` to `runApp`.
```

    ```dart
    void main(List<String> arguments) {
      if (arguments.isNotEmpty && arguments.first == 'version') {
        print('Dartpedia CLI version $version');
      } else if (arguments.isNotEmpty && arguments.first == 'help') {
        printUsage();
      } else if (arguments.isNotEmpty && arguments.first == 'search') { // Changed to 'search'
        // Modify this block:
        final inputArgs = arguments.length > 1 ? arguments.sublist(1) : null;
        runApp(inputArgs);
      } else {
        printUsage();
      }
    }
    ```

>     * **List manipulation:** `arguments.sublist(1)` creates a new list

Reviewer: I think you should preface these bulleted lists with something like this:

Highlights from the preceding code:

>     * **List manipulation:** `arguments.sublist(1)` creates a new list

Reviewer: ```suggestion
    * `arguments.sublist(1)` creates a new list
```

        containing all elements of the `arguments` list *after* the first
        element (which was `search`).

>     * **Conditional assignment to `null`:** The `arguments.length > 1 ? ... : null;`

Reviewer: ```suggestion
    * `arguments.length > 1 ? ... : null;`
```

        is a conditional (ternary) operator. It ensures that if no arguments

>         are provided after `search`, `inputArgs` becomes `null`, matching the

Reviewer: ```suggestion
        are provided after the `search` command, `inputArgs` becomes `null`, matching the
```

        sample code's behavior for `runApp`'s `arguments` parameter.

> 1.  **Test `runApp` with arguments:** Run the application with a test article

Reviewer: ```suggestion
1.  **Test `runApp` with arguments:** Using the command line, run the application with a test article
```

    title:

    ```bash
    dart bin/cli.dart search Dart Programming
    ```

    You should see:

    ```bash
    runApp received arguments: [Dart, Programming]
    ```

>     Run without extra arguments:

Reviewer: ```suggestion
    Next, run the same command without the extra arguments:
```

    ```bash
    dart bin/cli.dart search
    ```

    You should see:

    ```bash
    runApp received arguments: null
    ```

> 1.  **Handle missing article title and user input with `stdin`:** It's more

Reviewer: ```suggestion
1.  **Handle the missing article title and user input with the `stdin` command:** It's more
```

    user-friendly to prompt the user if they don't provide an article title on

>     the command line. We'll use `stdin.readLineSync()` for this.

Reviewer: ```suggestion
    the command line. Use `stdin.readLineSync()` for this.
```

        return; // Exit the function immediately after reading input from stdin
      } else {
        // We'll handle command-line arguments and their simulation here later.
        articleTitle = arguments.join(' '); // Join arguments into a single string
      }

      // We'll add the "Looking up..." and final simulation here later.
      print('Current article title: $articleTitle'); // This line will only run if arguments were provided
    }
    ```

    * `stdin.readLineSync()`: This function reads a line of text typed by the
        user into the console. Its return type is `String?`.
    * The `if (inputFromStdin == null || inputFromStdin.isEmpty)` performs a
        null check and an empty string check. If either is true, the program
        prints a message and `return`s, exiting the function.
    * `arguments.join(' ')`: This `List` method concatenates all elements of
        the `arguments` list into a single string, using a space as the
        separator. For example, `['Dart', 'Programming']` becomes
        `"Dart Programming"`. This is crucial for treating multi-word
        command-line inputs as a single search phrase.

> 2.  **Finalize `runApp` with single article simulation:** Now, let's complete

Author: Ok, I took a stab at this.

syntax.

:::secondary What you'll learn

* Implement basic control flow with `if/else`.
* Work with collections, specifically `List`s, and perform common operations
  like checking if a list is empty.
* Declare and use variables with `const` and `late String?`.
* Handle nullability with null checks.
* Define and call functions.
* Use string interpolation for dynamic text.
* Read user input from the command line using `stdin`.

:::

## Prerequisites

Before you begin this chapter, ensure you have:

* Completed Chapter 1 and have a working Dart development environment.
* Familiarity with basic programming concepts (variables, data types, control
  flow).

## Tasks

Start building your **Dartpedia** command-line application by adding some basic
functionality and exploring Dart syntax.

### Task 1: Implement version and help commands

1.  **Implement the `version` command in `cli/bin/cli.dart`:** Add logic to
    handle a `version` command. Use an `if` statement to check if the first
    argument provided is `version`. You'll also need a `version` constant.

    First, above your `main` function, declare a `const` variable for the
    version:

    ```dart

>     const version = '0.0.1'; // Add this line

Reviewer 2: I think there should be 1 sentence about what `const` means in the simplest terms possible (i.e. "it can never be changed")

        printUsage(); // Change this from 'Hello, Dart!'
      }
    }
    ```

5.  **Understand the `if/else` structure and variables:** You've now built an
    `if`, `else if`, and `else` structure for conditional execution.

    * `arguments.isNotEmpty` checks if any command-line arguments were
        provided.
    * `arguments.first` accesses the very first argument, which we're using as
        our command.
    * Notice the `version` variable is declared as a `const`. This means its
        value is known at compile time and you cannot change it during runtime.
    * `arguments`, on the other hand, is a regular (non-constant) variable
        because its content can change based on user input.

    Test the `help` command: Run your application with the help argument:

    ```bash
    dart bin/cli.dart help
    ```

    You should see the usage information printed.

    Also, try running it without any arguments:

    ```bash
    dart bin/cli.dart
    ```

    It should now also display the usage information.

    :::note
    Any command-line argument that is not `version` or `help` will currently
    also print the usage information. This is expected behavior for now.

>     :::

Reviewer 2: `help` does print usage info

Author: Good point!

### Task 2: Implement the search command

Next, implement a basic `search` command that takes an article title as
input. As you build this functionality, you'll work with `List` manipulation,
null checks, and string interpolation.

1.  **Integrate the `search` command into `main`:** First, modify your `main`
    function in `cli/bin/cli.dart` to include an `else if` branch that handles
    the `search` command. For now, we'll just print a placeholder message.

    ```dart
    void main(List<String> arguments) {
      if (arguments.isNotEmpty && arguments.first == 'version') {
        print('Dartpedia CLI version $version');
      } else if (arguments.isNotEmpty && arguments.first == 'help') {
        printUsage();
      } else if (arguments.isNotEmpty && arguments.first == 'search') {
        // Add this new block:
        print('Search command recognized!');
      } else {
        printUsage();
      }
    }
    ```

2.  **Test the new command:** Run your application with the `search`
    command:

    ```bash
    dart bin/cli.dart search
    ```

    You should see:

    ```bash
    Search command recognized!
    ```

> 3.  **Define the `runApp` function:** Your `search` command will eventually run
>     the core logic of our application. Next, create a new function called
>     `runApp` that will house this logic. Initially, `runApp` will just take a
>     list of arguments and print them. Place this new function below `main`.

Reviewer 2: ```suggestion
3.  **Define the `runApp` function:** The `search` command will eventually run
    the core logic of your application by calling a function called `runApp`. For now, 
    it will print the arguments passed into search. Place this new function below `main`.
```

        articleTitle = stdin.readLineSync();
        return; // Exits here if input is from stdin
      } else {
        articleTitle = arguments.join(' ');
      }

      // This section runs only if arguments were originally provided (not from stdin)
      print('Looking up articles about "$articleTitle". Please wait.'); // Using string interpolation
      print('Here ya go!');
      print('(Pretend this is an article about "$articleTitle")'); // Simulates finding the combined article
    }
    ```

9.  **Final Test Run with both scenarios:**

    Run with arguments (note the quotes around the search term in the output):

    ```bash
    dart bin/cli.dart search Dart Programming
    ```

    You should see:

    ```bash
    Looking up articles about "Dart Programming". Please wait.
    Here ya go!
    (Pretend this is an article about "Dart Programming")
    ```

    Run without arguments (type "Flutter Framework" when prompted):

    ```bash
    dart bin/cli.dart search
    ```

    ```bash
    Please provide an article title.
    Flutter Framework
    ```
    
    :::note
    The program will exit after you type "Flutter Framework" because `runApp`

>     returns immediately after `stdin` input as per the sample code's behavior.

Reviewer 2: This is the behavior that shouldn't have existed in the first place, I just looked at the step-by-step code and I have no idea why I did this

Author: I'm finding it harder than I expected to take a "snapshot" in my brain of the various changes made, especially as we make them in multiple places across multiple files.

I'm going to put together a sort of "roll-up" doc that shows, at the end of each chapter, files changed, and the state they are in. I know the step-by-step already does this, but it's a bit harder to peruse in the full file structure.

---
title: Making Your CLI Program Interactive
description: Add simple commands to your cli program. Learn the fundamentals of Dart syntax including control flow, collections, variables, functions, and more.
---

> In this chapter, you'll get hands-on with Dart syntax. You'll cover control

Reviewer 3: ```suggestion
In this chapter, you'll get hands-on practice with Dart syntax. You'll cover control
```

flow, collections, variables, functions, and more. You'll learn how to read user
input, print usage information, and create a basic command-line interaction. By

> the end, you'll have a foundational understanding of Dart syntax.

Reviewer 3: Perhaps a "What you'll learn" label with a bulleted list?

## Prerequisites

> 
> * Completed Chapter 1 and have a working Dart development environment.
> * Familiarity with basic programming concepts (variables, data types, control
>   low).

Reviewer 3: ```suggestion

Before working through this section, we suggest that you first:

* Complete Chapter 1 and have a working Dart development environment.
* Have a familiarity with basic programming concepts 
  (such as variables, data types, control flow).
```

Author: I think we should be stronger here, because the code from previous chapters is required to continue in the next chapters.

How about "Before you begin this chapter, ensure you have:"

## Tasks

Start building your Wikipedia command-line application by adding some basic

> functionality and exploring Dart syntax.

Reviewer 3: Again, I'd name this example something else. Even "Dictionary" is better, but I'm sure you can come up with something even better. :)

Author: I'm working with dartpedia for now, unless our super official chat-room vote decides otherwise.

> ### Task 1: Implement `Version` and `Help` Commands

Reviewer 3: ```suggestion
### Task 1: Implement `Version` and `Help` commands
```

I'm of a mixed mind here. Google developer doc guidelines says not to use code font in titles/anchors, but sometimes it clarifies...

Author: I removed them. I think the formatting throughout the rest of the docs will be helpful enough, and this way we can stay consistent.

    Update your `runApp` function. Notice that `stdin.readLineSync()` can return
    `null`, so we declare `articleTitle` as `late String?`. You'll need to
    import `dart:io` at the top of your `cli.dart` file.

    ```dart
    // Add this line at the top of cli/bin/cli.dart:
    import 'dart:io';

    // ... (your existing main function and const version)

    void runApp(List<String>? arguments) { // Add ? to arguments type
      late String? articleTitle; // Declare articleTitle with late and ?

      if (arguments == null || arguments.isEmpty) {
        print('Please provide an article title.');
        articleTitle = stdin.readLineSync(); // Read input from user
        if (articleTitle == null || articleTitle.isEmpty) { // Add basic check for empty input
          print('No article title provided. Exiting.');
          return; // Exit if no input
        }
      } else {
        articleTitle = arguments.join(', '); // Join command-line arguments
      }

      // We'll add more print statements here later
      print('Looking up articles about $articleTitle. Please wait.');
    }

    // ... (your existing printUsage() function)
    ```

>     *   **Null Safety:** `List<String>? arguments` means that the `arguments`

Reviewer 3: I'd be clearer that this can be null _because_ of the '?' notation.

      the input came from `stdin`, this list will contain just the single
      `articleTitle`. If it came from the command line, it will contain the
      `arguments` themselves.
    * **String Interpolation:** The
      `print('(Pretend this is an article about $arg)');` line uses string
      interpolation. The `$` sign allows you to embed variables directly into a
      string, making it easy to create dynamic output.

1.  **Final Test Run with both scenarios:**

    Run with arguments:

    ```bash
    dart bin/cli.dart wikipedia Dart Programming
    ```

    You should see:

    ```bash
    Looking up articles about Dart, Programming. Please wait.
    Here ya go!
    (Pretend this is an article about Dart)
    Here ya go!
    (Pretend this is an article about Programming)
    ```

    Run without arguments (type "Flutter" when prompted):

    ```bash
    dart bin/cli.dart wikipedia

    Please provide an article title.
    Flutter
    Looking up articles about Flutter. Please wait.
    Here ya go!
    (Pretend this is an article about Flutter)
    ```

    You have now successfully built the basic `wikipedia` command with user input
    handling.

> ### Task 3: Handle Null Input from `stdin`

Reviewer 3: ```suggestion
### Task 3: Handle null input from `stdin`
```

Just fix this globally.  :)

Author: Done
